### PR TITLE
[DNM] add co-processor client

### PR DESCRIPTION
**Pull request by disksing**

### CODE REVIEW

**Comment by disksing** [comment]
PTAL @qiuyesuifeng @ngaut @coocood @shenli

File: coprocessor.go [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+		}
+	}
+	return tasks, nil
+}
+
+type copIterator struct {
+	store *tikvStore
+	req   *kv.Request
+	tasks []copTask
+	index int
```
Comment (ngaut) [inline coprocessor.go]: Please add comments for index and tasks

File: coprocessor_test.go [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+	// nil --- 'g' --- 'n' --- 't' --- nil
+	// <-  1  -> <- 2 -> <- 3 -> <- 4 ->
+	pd.setStore(100, "addr100")
+	pd.setRegion(1, nil, []byte("g"), []uint64{100})
+	pd.setRegion(2, []byte("g"), []byte("n"), []uint64{100})
+	pd.setRegion(3, []byte("n"), []byte("t"), []uint64{100})
+	pd.setRegion(4, []byte("t"), nil, []uint64{100})
+
+	tasks, err := buildRegionTasks(cache, s.buildKeyRanges("a", "c"))
```
Comment (shenli) [inline coprocessor_test.go]: Add some cases that contains g/n/t/nil?

File: coprocessor.go [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+		index: index,
+	}
+}
+
+type copTask struct {
+	region *Region
+	ran    kv.KeyRange
+}
+
+func buildRegionTasks(cache *RegionCache, ranges []kv.KeyRange) ([]copTask, error) {
```
Comment (coocood) [inline coprocessor.go]: A task should be able to contain multiple ranges.

File: client.go [hunk 4]
```diff
@@ -71,15 +71,53 @@ func NewClient(srvHost string) (*Client, error) {
 	}, nil
 }
 
-// Send send a Request and receive Response.
-func (c *Client) Send(req *pb.Request) (*pb.Response, error) {
+// SendCopReq sends a Request to co-processor and receive Response.
```
Comment (shenli) [inline client.go]: receives

File: coprocessor.go [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+				},
+			}
+			if region.Contains(r.EndKey) {
+				// the whole range is in region
+				tasks = append(tasks, task)
+				break
+			}
+			task.ran.EndKey = region.EndKey()
+			tasks = append(tasks, task)
+			r.StartKey = region.EndKey()
```
Comment (coocood) [inline coprocessor.go]: Why change range startKey?

File: client.go [hunk 6]
```diff
@@ -71,15 +71,53 @@ func NewClient(srvHost string) (*Client, error) {
...
+		// This connection is not valid any more, so close its underlying conn.
+		if connPool, ok := conn.(*pool.PoolConn); ok {
+			connPool.MarkUnusable()
+		}
+		return nil, errRetryable
+	}
+	return msg.GetCopResp(), nil
+}
+
+// SendKVReq send a Request to kv server and receive Response.
```
Comment (shenli) [inline client.go]: sends/reveives

File: client.go [hunk 7]
```diff
@@ -71,15 +71,53 @@ func NewClient(srvHost string) (*Client, error) {
...
+	err = c.doSend(conn, &msg)
+	if err == nil {
+		if msg.GetMsgType() != msgpb.MessageType_CopReq || msg.GetCopResp() == nil {
+			err = errors.Trace(errInvalidResponse)
+		}
+	}
+	if err != nil {
+		log.Warn(err)
+		// This connection is not valid any more, so close its underlying conn.
+		if connPool, ok := conn.(*pool.PoolConn); ok {
```
Comment (shenli) [inline client.go]: connPool->c/poolConn

**Comment by ngaut** [comment]
@coocood PTAL

File: coprocessor.go [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+	}
+}
+
+type copTask struct {
+	region *Region
+	ran    kv.KeyRange
+}
+
+func buildRegionTasks(cache *RegionCache, ranges []kv.KeyRange) ([]copTask, error) {
+	var tasks []copTask
```
Comment (shenli) [inline coprocessor.go]: Why not use []*copTask?

**Comment by disksing** [comment]
@coocood @shenli  PTAL

**Comment by shenli** [comment]
LGTM

File: coprocessor.go [hunk 9]
```diff
@@ -63,44 +63,47 @@ type copTask struct {
...
-					StartKey: r.StartKey,
-				},
-			}
-			tasks = append(tasks, last)
+func appendTask(tasks []*copTask, cache *RegionCache, r kv.KeyRange) ([]*copTask, error) {
+	var last *copTask
+	if len(tasks) > 0 {
+		last = tasks[len(tasks)-1]
+	}
+	// Ensure `r.StartKey` is inside `last`, create task if need.
```
Comment (coocood) [inline coprocessor.go]: There should be a for loop.

Comment (disksing) [inline coprocessor.go]: @coocood  There is a for loop at L67 for iteration.

File: coprocessor.go [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+		// The whole range is inside last task.
+		last.ran.EndKey = r.EndKey
+	} else {
+		// Part of r is not in the range of last task.
+		last.ran.EndKey = last.region.EndKey()
+		remain := kv.KeyRange{
+			StartKey: last.region.EndKey(),
+			EndKey:   r.EndKey,
+		}
+		return appendTask(tasks, cache, remain)
```
Comment (shenli) [inline coprocessor.go]: @coocood This recursion saves the loop.

File: coprocessor.go (hunk 9 above)
Comment (coocood) [inline coprocessor.go]: Got it

File: coprocessor_test.go [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+	s.taskEqual(c, tasks[1], 3, "o", "p")
+
+	tasks, err = buildCopTasks(cache, s.buildKeyRanges("h", "k", "m", "p"))
+	c.Assert(err, IsNil)
+	c.Assert(tasks, HasLen, 2)
+	s.taskEqual(c, tasks[0], 2, "h", "n")
+	s.taskEqual(c, tasks[1], 3, "n", "p")
+}
+
+func (s *testCoprocessorSuite) buildKeyRanges(keys ...string) []kv.KeyRange {
```
Comment (coocood) [inline coprocessor_test.go]: Key ranges are not continuous.
key ranges should not be built this way.

File: client.go [hunk 12]
```diff
@@ -71,24 +71,62 @@ func NewClient(srvHost string) (*Client, error) {
...
+		if msg.GetMsgType() != msgpb.MessageType_KvResp || msg.GetKvResp() == nil {
+			err = errors.Trace(errInvalidResponse)
+		}
+	}
 	if err != nil {
 		log.Warn(err)
 		// This connection is not valid any more, so close its underlying conn.
-		if connPool, ok := conn.(*pool.PoolConn); ok {
-			connPool.MarkUnusable()
+		if pollConn, ok := conn.(*pool.PoolConn); ok {
```
Comment (zimulala) [inline client.go]: pollConn -> poolConn ?

Comment (disksing) [inline client.go]: @zimulala Thanks!

File: coprocessor_test.go (hunk 11 above)
Comment (disksing) [inline coprocessor_test.go]: @coocood The ranges could be discrete. Parameters come in pairs (start, end, start, end...)

File: coprocessor.go [hunk 13]
```diff
@@ -0,0 +1,168 @@
...
+// copTask contains a related Region and KeyRange for a kv.Request.
+type copTask struct {
+	region *Region
+	ran    kv.KeyRange
+}
+
+func buildCopTasks(cache *RegionCache, ranges []kv.KeyRange) ([]*copTask, error) {
+	var tasks []*copTask
+	for _, r := range ranges {
+		var err error
```
Comment (zimulala) [inline coprocessor.go]: Move line 68 code behind the line 66 is better ?

**Comment by coocood** [comment]
LGTM

File: coprocessor.go (hunk 13 above)
Comment (disksing) [inline coprocessor.go]: @zimulala I think these two approaches are basically the same at runtime, it won't lead to more memory allocation.

**Comment by zimulala** [comment]
LGTM

**Pull request merged**